### PR TITLE
Iss hipm 706

### PR DESCRIPTION
**Pull request by Senn3**
Added a page where the user can see the ratings of an exhibit and is able to send his own rating to the server.

### CODE REVIEW

**Comment by cbruegg** [comment]
Could you please merge in the changes from `master`? I'll review afterwards.

**Comment by cbruegg** [comment]
Looks like the changes break something:

`2017-12-03T17:47:42.9015073Z ##[error]Sources\HipMobileUI\Pages\LoadingPage.xaml(53,26): Error : Position 53:26. Can not set the content of controls1:FadeInImage as it doesn't have a ContentPropertyAttribute`

File: Sources/HiPMobile/BusinessLayer/Managers/UserRatingManager.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+using PaderbornUniversity.SILab.Hip.Mobile.Shared.ServiceAccessLayer.ContentApiAccesses;
+using PaderbornUniversity.SILab.Hip.Mobile.Shared.ServiceAccessLayer.ContentApiAccesses.Contracts;
+using System.Threading.Tasks;
+
+namespace PaderbornUniversity.SILab.Hip.Mobile.Shared.BusinessLayer.Managers
+{
+
+    public interface IUserRatingManager
+    {
+        Task<UserRating> GetUserRating(Exhibit exhibit);
```
Comment (cbruegg) [inline Sources/HiPMobile/BusinessLayer/Managers/UserRatingManager.cs]: Methods returning `Task` should have the suffix `Async`, e.g. `GetUserRatingAsync`. Also applies to other newly added methods.

File: Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        public UserRating(UserRatingDto userRatingDto)
+        {
+            Id = userRatingDto.Id;
+            Average = userRatingDto.Average;
+            Count = userRatingDto.Count;
+            RatingTable = userRatingDto.RatingTable;
+        }
+
+        public int Id { get; set; }
```
Comment (cbruegg) [inline Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs]: Why do all these properties have setters? This class should be easy to make immutable.

File: Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        public double Average { get; set; }
+
+        public int Count { get; set; }
+
+        public Dictionary<int, int> RatingTable { get; set; }
```
Comment (cbruegg) [inline Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs]: What does this dictionary contain? Can you add documentation?

File: Sources/HiPMobile/HiPMobile.csproj [hunk 4]
```diff
@@ -274,6 +279,15 @@
     <Reference Include="System.Net.Http.Primitives, Version=4.2.29.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
       <HintPath>..\..\packages\Microsoft.Net.Http.2.2.29\lib\portable-net45+win8\System.Net.Http.Primitives.dll</HintPath>
     </Reference>
+    <Reference Include="Xamarin.Forms.Core, Version=2.0.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (cbruegg) [inline Sources/HiPMobile/HiPMobile.csproj]: `HipMobile` should not have dependencies on UI layer libraries. For what exactly are these three new references needed in this particular project?

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            try
+            {
+                string url = RequestPath + exhibit.IdForRestApi;
+                var json = await contentApiClient.GetResponseFromUrlAsString(url);
+                return JsonConvert.DeserializeObject<UserRatingDto>(json);
+            }
+            catch (NotFoundException)
+            {
+                return new UserRatingDto();
```
Comment (cbruegg) [inline Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: Is this a good idea? Is this a situation that can actually legally happen?

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        public UserRatingApiAccess(IContentApiClient contentApiClient)
+        {
+            this.contentApiClient = contentApiClient;
+        }
+
+        public async Task<UserRatingDto> GetUserRating(Exhibit exhibit)
+        {
+            try
+            {
+                string url = RequestPath + exhibit.IdForRestApi;
```
Comment (cbruegg) [inline Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: Should be a `var`

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            }
+            catch (NotFoundException)
+            {
+                return new UserRatingDto();
+            }
+        }
+
+        public async Task<bool> SendUserRating(Exhibit exhibit, int rating)
+        {
+            string url = RequestPath + exhibit.IdForRestApi + "?Rating=" + rating;
```
Comment (cbruegg) [inline Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: `var` + string interpolation (`var url = $"{RequestPath}{exhibit.IdForRestApi}[...]`) would be nice here

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            var result = await contentApiClient.PostRequestBody(url, String.Empty);
+            if (result.StatusCode == HttpStatusCode.Created)
+            {
+                string jsonPayload = await result.Content.ReadAsStringAsync();
```
Comment (cbruegg) [inline Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: `var`

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/Contracts/IUserRatingApiAccess.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+using System.Threading.Tasks;
+using PaderbornUniversity.SILab.Hip.Mobile.Shared.ServiceAccessLayer.ContentApiDtos;
+using PaderbornUniversity.SILab.Hip.Mobile.Shared.BusinessLayer.Models;
+
+namespace PaderbornUniversity.SILab.Hip.Mobile.Shared.ServiceAccessLayer.ContentApiAccesses.Contracts
+{
+    public interface IUserRatingApiAccess
+    {
+        Task<UserRatingDto> GetUserRating(Exhibit exhibit);
+        Task<bool> SendUserRating(Exhibit exhibit, int rating);
```
Comment (cbruegg) [inline Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/Contracts/IUserRatingApiAccess.cs]: Could you please document what the return value means?

File: Sources/HiPMobile/packages.config [hunk 10]
```diff
@@ -32,4 +32,5 @@
   <package id="System.Threading" version="4.3.0" targetFramework="portable45-net45+win8" />
   <package id="Unity" version="4.0.1" targetFramework="portable45-net45+win8" />
   <package id="Xam.Plugins.Settings" version="2.5.1.0" targetFramework="portable45-net45+win8" />
+  <package id="Xamarin.Forms" version="2.4.0.280" targetFramework="portable45-net45+win8" />
```
Comment (cbruegg) [inline Sources/HiPMobile/packages.config]: As noted above already, is this needed?

File: Sources/HipMobileUI.Droid/HipMobileUI.Droid.csproj [hunk 11]
```diff
@@ -254,22 +255,22 @@
...
-      <HintPath>..\..\packages\Xamarin.Forms.2.3.4.247\lib\MonoAndroid10\Xamarin.Forms.Platform.dll</HintPath>
+      <HintPath>..\..\packages\Xamarin.Forms.2.4.0.280\lib\MonoAndroid10\Xamarin.Forms.Platform.dll</HintPath>
     </Reference>
     <Reference Include="Xamarin.Forms.Platform.Android, Version=2.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Xamarin.Forms.2.3.4.247\lib\MonoAndroid10\Xamarin.Forms.Platform.Android.dll</HintPath>
+      <HintPath>..\..\packages\Xamarin.Forms.2.4.0.280\lib\MonoAndroid10\Xamarin.Forms.Platform.Android.dll</HintPath>
     </Reference>
     <Reference Include="Xamarin.Forms.Xaml, Version=2.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Xamarin.Forms.2.3.4.247\lib\MonoAndroid10\Xamarin.Forms.Xaml.dll</HintPath>
+      <HintPath>..\..\packages\Xamarin.Forms.2.4.0.280\lib\MonoAndroid10\Xamarin.Forms.Xaml.dll</HintPath>
```
Comment (cbruegg) [inline Sources/HipMobileUI.Droid/HipMobileUI.Droid.csproj]: Thanks for updating this!

File: Sources/HipMobileUI.Droid/app.config [hunk 12]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (cbruegg) [inline Sources/HipMobileUI.Droid/app.config]: What is this file needed for?

File: Sources/HipMobileUI.iOS/app.config [hunk 13]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (cbruegg) [inline Sources/HipMobileUI.iOS/app.config]: See comment on other `app.config` above.

File: Sources/HipMobileUI/Helpers/Settings.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+using Plugin.Settings.Abstractions;
+
+namespace PaderbornUniversity.SILab.Hip.Mobile.UI.Helpers
+{
+  /// <summary>
+  /// This is the Settings static class that can be used in your Core solution or in any
+  /// of your client applications. All settings are laid out the same exact way with getters
+  /// and setters. 
+  /// </summary>
+  public static class Settings
```
Comment (cbruegg) [inline Sources/HipMobileUI/Helpers/Settings.cs]: We already have such a class somewhere in our codebase, why did you add this here?

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+-->
+
+<ContentPage xmlns="http://xamarin.com/schemas/2014/forms" xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             x:Class="PaderbornUniversity.SILab.Hip.Mobile.UI.Pages.UserRatingPage"
+             xmlns:controls1="clr-namespace:PaderbornUniversity.SILab.Hip.Mobile.UI.Controls;assembly=HipMobileUI"
+             xmlns:behaviors="clr-namespace:PaderbornUniversity.SILab.Hip.Mobile.UI.Behaviors;assembly=HipMobileUI"
+             xmlns:helpers="clr-namespace:PaderbornUniversity.SILab.Hip.Mobile.UI.Helpers;assembly=HipMobileUI"
+             Title="{Binding Headline}">
+    <ContentPage.Content>
+        <AbsoluteLayout >
```
Comment (cbruegg) [inline Sources/HipMobileUI/Pages/UserRatingPage.xaml]: Nitpick: There's a space before `>`

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+// limitations under the License.
+using PaderbornUniversity.SILab.Hip.Mobile.UI.Navigation;
+using PaderbornUniversity.SILab.Hip.Mobile.UI.ViewModels.Pages;
+
+
+namespace PaderbornUniversity.SILab.Hip.Mobile.UI.Pages 
+{
+    public partial class UserRatingPage : IViewFor<UserRatingPageViewModel> 
+	{
+        public UserRatingPage()
```
Comment (cbruegg) [inline Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs]: Wrong formatting

File: Sources/HipMobileUI/Resources/Strings.resx [hunk 17]
```diff
@@ -822,10 +822,55 @@ Verantwortlich für den Betrieb: Dr. Simon Oberthür</value>
...
+    <value>Sie müssen sich erst anmelden oder registrieren, bevor eine Bewertung abgegeben werden kann.</value>
+  </data>
+  <data name="UserRating_Dialog_Message_No_Internet" xml:space="preserve">
+    <value>Es kann keine Verbindung zum Internet aufgebaut werden.</value>
+  </data>
+  <data name="UserRating_Dialog_Message_No_Rating" xml:space="preserve">
+    <value>Es wurde keiner der Sterne zum Bewerten markiert.</value>
+  </data>
+  <data name="UserRating_Dialog_Title_Not_Logged_In" xml:space="preserve">
+    <value>Nicht Angemeldet</value>
```
Comment (cbruegg) [inline Sources/HipMobileUI/Resources/Strings.resx]: _Angemeldet_ shouldn't be capitalized

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs [hunk 18]
```diff
@@ -91,6 +100,46 @@ public AppetizerPageViewModel(Exhibit exhibit)
             await Navigation.PushAsync(downloadPage);
         }
 
+        /// <summary>
+        /// Sets the average user rating, the count and the stars if the user is connected to the internet.
+        /// </summary>
+        /// <returns></returns>
+        private async void SetUserRating()
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: Maybe call this `RefreshRating` to make more clear that this method does something on the network? Just an idea.

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs [hunk 19]
```diff
@@ -91,6 +100,46 @@ public AppetizerPageViewModel(Exhibit exhibit)
...
+        {
+            if (IoCManager.Resolve<INetworkAccessChecker>().GetNetworkAccessStatus() == NetworkAccessStatus.NoAccess)
+            {
+                RatingAverage = "-";
+                RatingStars = "?????";
+                RatingCount = "- " + Strings.UserRating_Rate_Count;
+            }
+            else
+            {
+                UserRating userRating = await IoCManager.Resolve<IUserRatingManager>().GetUserRating(exhibit);
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: `var`

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs [hunk 20]
```diff
@@ -91,6 +100,46 @@ public AppetizerPageViewModel(Exhibit exhibit)
...
+                string stars = "";
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: `var`

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs [hunk 21]
```diff
@@ -91,6 +100,46 @@ public AppetizerPageViewModel(Exhibit exhibit)
...
+                else
+                    RatingAverage = "-";
+                RatingStars = stars;
+                RatingCount = userRating.Count.ToString() + " " + Strings.UserRating_Rate_Count;
+            }
+        }
+
+        private async void GoToUserRatingPage()
+        {
+            if (IoCManager.Resolve<INetworkAccessChecker>().GetNetworkAccessStatus() != NetworkAccessStatus.NoAccess)
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: And else? Maybe let the user know.

File: Sources/HipMobileUI/ViewModels/Pages/ExhibitDetailsViewModel.cs [hunk 22]
```diff
@@ -207,10 +208,24 @@ private void ToggleVisibilityOfNavigationButtons()
...
                 currentViewIndex++;
-                NextViewAvailable = currentViewIndex < pages.Count - 1;
+                NextViewAvailable = currentViewIndex < pages.Count;
                 PreviousViewAvailable = true;
                 await SetCurrentView();
             }
+            else if (currentViewIndex == pages.Count - 1)
+            {
+                // stop audio
+                if (IoCManager.Resolve<INetworkAccessChecker>().GetNetworkAccessStatus() != NetworkAccessStatus.NoAccess)
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/ExhibitDetailsViewModel.cs]: As above, what else?

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+using System.IO;
+using PaderbornUniversity.SILab.Hip.Mobile.Shared.BusinessLayer.DtoToModelConverters;
+using System.Collections.Generic;
+
+namespace PaderbornUniversity.SILab.Hip.Mobile.UI.ViewModels.Pages
+{
+    public class UserRatingPageViewModel : NavigationViewModel
+    {
+
+        #region Variabels
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: Typo

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+        private string star5BarCount;
+
+        private string ratingAverage;
+        private string ratingStars;
+        private string ratingCount;
+
+        private int rating = 0;
+        #endregion
+
+
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: I think this blank line is too much

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 25]
```diff
@@ -0,0 +1,369 @@
...
+            var imageData = exhibit.Image.GetDataBlocking();
+            Image = imageData != null ? ImageSource.FromStream(() => new MemoryStream(imageData)) : ImageSource.FromStream(() => new MemoryStream(BackupData.BackupImageData));
+
+            SetUserRatingUi();
+            SendRatingCommand = new Command(SendUserRating);
+            SelectStarCommand = new Command(OnSelectStar);
+
+        }
+        private async void SetUserRatingUi()
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: There should be a blank line before this. Also again, is `Set[...]` the best name for this?

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 26]
```diff
@@ -0,0 +1,369 @@
...
+        private int rating = 0;
+        #endregion
+
+
+        public UserRatingPageViewModel(Exhibit exhibit)
+        {
+            Exhibit = exhibit;
+            Headline = exhibit.Name;
+
+            var imageData = exhibit.Image.GetDataBlocking();
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: This method is obsolete and should only be used in legacy code. Please use the asynchronous version.

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 27]
```diff
@@ -0,0 +1,369 @@
...
+        {
+            UserRating userRating = await IoCManager.Resolve<IUserRatingManager>().GetUserRating(exhibit);
+            SetRatingAverageAndCount(userRating.Average, userRating.Count);
+            SetRatingBars(userRating.RatingTable, userRating.Count);
+            SetRatingStars();
+        }
+
+        private void SetRatingAverageAndCount(double average, int count)
+        {
+            string stars = "";
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: `var`

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 28]
```diff
@@ -0,0 +1,369 @@
...
+        private GridLength CalculateBarProportion(double value, double totalCount)
+        {
+            if (value == 0)
+                return new GridLength(0, GridUnitType.Absolute);
+            if (value == totalCount)
+                return new GridLength(short.MaxValue, GridUnitType.Star);
+            double prop = (1 / (1 - value / totalCount) - 1);
+            return new GridLength(prop, GridUnitType.Star);
+        }
+
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: Here and after other methods are extra blank lines. There should only be one blank line in between class members.

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs [hunk 29]
```diff
@@ -0,0 +1,369 @@
...
+                    Star1 = "★";
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Sends the user rating to the server if the following conditions are given:
+        /// 1. The user is connected to the internet
+        /// 2. The user is logged in
+        /// 3. The user haas selected a star
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: Typo

File: Sources/HipMobileUI/ViewModels/Views/UserRatingListItemViewModel.cs [hunk 30]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        private string text;
+        public string Text {
+            get { return text; }
+            set { SetProperty(ref text, value); }
+        }
+
+        public UserRatingListItemViewModel(Exhibit exhibit) {
+            UserName = "Name";
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Views/UserRatingListItemViewModel.cs]: I don't think there should be sample data here.

File: Sources/HiPMobile/BusinessLayer/Managers/UserRatingManager.cs (hunk 1 above)
Comment (Senn3) [reply Sources/HiPMobile/BusinessLayer/Managers/UserRatingManager.cs]: Added the suffix "Async" to all method with "Task" as return type

File: Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs (hunk 2 above)
Comment (Senn3) [reply Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs]: Ive removed the setters

File: Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs (hunk 3 above)
Comment (Senn3) [reply Sources/HiPMobile/BusinessLayer/Models/ModelClasses/UserRating.cs]: Added doc for the dictionary

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs (hunk 6 above)
Comment (Senn3) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: Changed to var

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs (hunk 8 above)
Comment (Senn3) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: Changed to var

File: Sources/HipMobileUI.Droid/app.config (hunk 12 above)
Comment (Senn3) [reply Sources/HipMobileUI.Droid/app.config]: I don't know. This file was automaticly created. I will remove this one and the other one (iOS/app.config)

File: Sources/HipMobileUI/Helpers/Settings.cs (hunk 14 above)
Comment (Senn3) [reply Sources/HipMobileUI/Helpers/Settings.cs]: This was a mistake. I will remove this file in the next commit

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml (hunk 15 above)
Comment (Senn3) [reply Sources/HipMobileUI/Pages/UserRatingPage.xaml]: Removed the space

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs (hunk 16 above)
Comment (Senn3) [reply Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs]: And where is the wrong formatting here? This file was generated automatically

File: Sources/HipMobileUI/Resources/Strings.resx (hunk 17 above)
Comment (Senn3) [reply Sources/HipMobileUI/Resources/Strings.resx]: Changed to "angemeldet"

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs (hunk 19 above)
Comment (Senn3) [reply Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: Changed to var

File: Sources/HipMobileUI/ViewModels/Views/UserRatingListItemViewModel.cs (hunk 30 above)
Comment (Senn3) [reply Sources/HipMobileUI/ViewModels/Views/UserRatingListItemViewModel.cs]: This file is obsolet anyway and will be removed the next commit

File: Sources/HiPMobile/packages.config (hunk 10 above)
Comment (Senn3) [reply Sources/HiPMobile/packages.config]: It doesnt seems to be needed. Removed in next commit

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs (hunk 5 above)
Comment (Senn3) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: Yes. If the server is not reachable or something with the link is wrong, a NotFoundException will be thrown and a unknown user rating is shown. It's also possible to return null and set the user rating to unknown somewhere else, e.g. in the user rating manager

Comment (cbruegg) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: If the server is not reachable, a `NotFoundException` shouldn't be thrown actually, if my memory serves me right. And if something with the link is wrong, then it makes sense to fail here instead of silently returning data the API actually never sent. `ApiAccess` classes should be dumb and simply act as an interface to the server, at least in my opinion. What do you think?

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs (hunk 16 above)
Comment (cbruegg) [reply Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs]: The indentation of the curly braces is wrong. You can use the autoformatter and see how it changes the file correctly.

**Comment by cbruegg** [comment]
One more thing, it would be great if you wrote a test for your `IUserRatingApiAccess` implementation similar to this: https://github.com/HiP-App/HiP-Forms/blob/master/Sources/HipMobile.Tests/ServiceAccessLayer/ContentApiAccesses/AchievementsApiAccessTest.cs

File: Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs (hunk 26 above)
Comment (Senn3) [reply Sources/HipMobileUI/ViewModels/Pages/UserRatingPageViewModel.cs]: Replaced with Image.GetDataAsync()

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs (hunk 21 above)
Comment (Senn3) [reply Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: Added a dialog for the user

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs (hunk 18 above)
Comment (Senn3) [reply Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: Changed the name to RefreshUserRating

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs (hunk 16 above)
Comment (Senn3) [reply Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs]: This seems to be a VS or github problem. I have looked into this file on my pc and the braces were formatted well.

File: Sources/HipMobileUI.Droid/HipMobileUI.Droid.csproj (hunk 11 above)
Comment (Senn3) [reply Sources/HipMobileUI.Droid/HipMobileUI.Droid.csproj]: You're welcome :)

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs (hunk 5 above)
Comment (Senn3) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: A NotFoundException is thrown when their is no user rating for the exhibit id on the server. 
For the moment I have removed the try catch block because of each exhibit a user rating should be existing, but it would be the best if we talk about this issue on monday.

File: Sources/HiPMobile/HiPMobile.csproj (hunk 4 above)
Comment (Senn3) [reply Sources/HiPMobile/HiPMobile.csproj]: Removed all Xamarin references from the HiPMobile project file

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/Contracts/IUserRatingApiAccess.cs (hunk 9 above)
Comment (cbruegg) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/Contracts/IUserRatingApiAccess.cs]: Have you seen this? GitHub has collapsed this comment because this line was changed.

File: Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs (hunk 5 above)
Comment (cbruegg) [reply Sources/HiPMobile/ServiceAccessLayer/ContentApiAccesses/UserRatingApiAccess.cs]: Yes, let's discuss on Monday.

File: Sources/HiPMobile/packages.config (hunk 10 above)
Comment (cbruegg) [reply Sources/HiPMobile/packages.config]: I still see it [here](https://github.com/HiP-App/HiP-Forms/blob/933ef384945ca586866a86cf02678f054b5eb485/Sources/HiPMobile/HiPMobile.csproj#L316)

File: Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs (hunk 16 above)
Comment (cbruegg) [reply Sources/HipMobileUI/Pages/UserRatingPage.xaml.cs]: The problem was that you mixed tabs and spaces in this file. I've quickly fixed it.

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs [hunk 31]
```diff
@@ -104,17 +109,23 @@ public AppetizerPageViewModel(Exhibit exhibit)
...
+                {
+                    Title = Strings.UserRating_Dialog_Title_No_Internet,
+                    Message = Strings.UserRating_Dialog_Message_No_Internet,
+                    OkText = Strings.UserRating_Ok
+                });
             }
             else
             {
-                var userRating = await IoCManager.Resolve<IUserRatingManager>().GetUserRatingAsync(exhibit);
+                var userRating = await IoCManager.Resolve<IUserRatingManager>().GetUserRatingAsync(exhibit.IdForRestApi);
```
Comment (cbruegg) [inline Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: This can throw in case the network is down, right? The check above with the `INetworkAccessChecker` doesn't ensure that the network will not go down shortly afterwards. You should wrap this in a try-catch with error handling, otherwise the app might crash on network errors.

File: Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs (hunk 21 above)
Comment (cbruegg) [reply Sources/HipMobileUI/ViewModels/Pages/AppetizerPageViewModel.cs]: I don't see it here, only in another place?

**Comment by cbruegg** [comment]
Most things seem to be fixed now :+1: Just to sum up what will still need to be dealt with:

- Merge in latest changes from `master`
- https://github.com/HiP-App/HiP-Forms/pull/153#discussion_r160022589
- https://github.com/HiP-App/HiP-Forms/pull/153#discussion_r160022645
- https://github.com/HiP-App/HiP-Forms/pull/153#discussion_r160022689

I've also now tested the changes on my device. Your code seems to work fine and the UI looks good, but for some reason the rating I send never makes it into the average. When I rate and come back, there's still no rating. I suspect that's the API's fault? If yes, did you let Team Fox know about this already?

An additional issue I've seen is that when the user has rated something and returns to the rating screen of that exhibit, no stars are selected and the user doesn't know whether he already rated this exhibit or not. Unfortunately it seems like the API provides no method to get the particular rating of one user, so maybe I can just open an issue on JIRA for this and we can improve this later?
